src/char/char.c

```diff
@@ -696,63 +696,53 @@ int char_mmo_char_tosql(int char_id, struct mmo_charstatus* p)
 }
 
 /**
- * Saves an array of 'item' entries into the specified table.
- *
- * @param items       The items array.
- * @param max         The array size.
- * @param id          The character/account/guild ID (depending on tableswitch).
- * @param tableswitch The type of table (@see enum inventory_table_type).
- * @return Error code.
- * @retval 0 in case of success.
+ * Gets a player object's item data from an sql table. [Smokexyz/Hercules]
+ * @param items     [in|out]   reference to the item list of a character/account/guild.
+ * @param max       [in]       Max amount of items to be pulled into the list.
+ * @param guid      [in]       Unique ID of the player object (account_id, char_id, guild_id).
+ * @param table     [in]       Table to be used for the transaction.
+ * @return -1 on failure or number of items added to the list if successful.
  */
-int char_memitemdata_to_sql(const struct item items[], int max, int id, int tableswitch)
+int char_getitemdata_from_sql(struct item *items, int max, int guid, enum inventory_table_type table)
 {
-	StringBuf buf;
+	int i = 0;
 	struct SqlStmt *stmt = NULL;
-	int i, j;
 	const char *tablename = NULL;
 	const char *selectoption = NULL;
 	bool has_favorite = false;
+	StringBuf buf;
 	struct item item = { 0 }; // temp storage variable
-	bool *flag = NULL; // bit array for inventory matching
-	bool found;
-	int errors = 0;
 
-	nullpo_ret(items);
+	nullpo_retr(-1, items);
+	Assert_retr(-1, guid > 0);
+	Assert_retr(-1, max > 0);
 
-	switch (tableswitch) {
+	switch (table) {
 	case TABLE_INVENTORY:     tablename = inventory_db;     selectoption = "char_id";    has_favorite = true; break;
 	case TABLE_CART:          tablename = cart_db;          selectoption = "char_id";    break;
-	case TABLE_STORAGE:       tablename = storage_db;       selectoption = "account_id"; break;
 	case TABLE_GUILD_STORAGE: tablename = guild_storage_db; selectoption = "guild_id";   break;
 	default:
-		ShowError("Invalid table name!\n");
-		Assert_retr(1, tableswitch);
+		ShowError("char_getitemdata_from_sql: Invalid table name!\n");
+		Assert_retr(-1, table);
```

Comment:
Well, let's put this in different terms. I'll expand the Assert_retr() macro in this code snippet:


```
enum inventory_table_type {
	TABLE_INVENTORY, // implicit 0
	TABLE_CART,
	TABLE_STORAGE,
	TABLE_GUILD_STORAGE,
};

switch (table) {
case TABLE_INVENTORY:
    tablename = inventory_db;
    selectoption = "char_id";
    has_favorite = true; break;
case TABLE_CART:
    tablename = cart_db;
    selectoption = "char_id";
    break;
case TABLE_GUILD_STORAGE:
    tablename = guild_storage_db;
    selectoption = "guild_id";
    break;
default:
    ShowError("char_getitemdata_from_sql: Invalid table name!\n");
    if (!table) {
        nullpo->assert_report(__FILE__, __LINE__, __func__, "table", "failed assertion");
        return(-1);
    }
 ```

Now, this is clearly not intended. The code returns only if `table` is zero. Now, a good static analyzer could detect the if as useless and warn about it. A good compiler could optimize it out. It's still wrong. And a change in the enum would subtly change the function behavior in a very hard to detect way (think what happens if one day we decide to add `TABLE_UNKNOWN = 0` at the top of the enum for example)